File: rust/08_stack/stack_based_on_linked_list.rs
#[derive(Hash, Eq, PartialEq, Debug, Default, Clone)]
pub struct ListNode {
    val: String,
    next: Option<Box<ListNode>>,
}

#[derive(Hash, Eq, PartialEq, Debug, Default, Clone)]
pub struct LinkedListStack {
    node: Option<Box<ListNode>>,
}

impl ListNode {
    fn new(val: String) -> Self {
        ListNode { val: val, next: None }
    }
}

impl LinkedListStack {
    pub fn new() -> Self {
        Default::default()
    }

    pub fn push(&mut self, x: String) {
        let mut n = ListNode::new(x);
        n.next = self.node.clone();
        self.node = Some(Box::new(n));
    }

    pub fn pop(&mut self) -> String {
        if self.is_empty() { return "-1".to_string(); }

        let val = self.node.as_ref().unwrap().val.clone();
        self.node = self.node.as_mut().unwrap().next.take();
        val.to_string()
    }

    pub fn print_all(&mut self) {
        let mut list = String::from("");

        while let Some(n) = self.node.as_mut() {
            list.push_str(&(n.val).to_string());
            list.push_str("-->");
            self.node = n.next.take();
        }
        println!("{:?}", list);
    }

    pub fn clear(&mut self) {
        self.node = None;
    }

    pub fn is_empty(&self) -> bool {
        if self.node.is_none() { true } else { false }
    }
}
//
// fn main() {
//     let mut stack = LinkedListStack::new();
//     stack.push("https://www.baidu.com".to_string());
//     stack.push("https://www.google.com".to_string());
//     stack.pop();
//     stack.push("https://twitter.com".to_string());
//     stack.print_all();
// }

File: go/08_stack/StackBasedOnLinkedList.go
package _8_stack

import "fmt"

/*
基于链表实现的栈
*/
type node struct {
	next *node
	val  interface{}
}

type LinkedListStack struct {
	//栈顶节点
	topNode *node
}

func NewLinkedListStack() *LinkedListStack {
	return &LinkedListStack{nil}
}

func (this *LinkedListStack) IsEmpty() bool {
	return this.topNode == nil
}

func (this *LinkedListStack) Push(v interface{}) {
	this.topNode = &node{next: this.topNode, val: v}
}

func (this *LinkedListStack) Pop() interface{} {
	if this.IsEmpty() {
		return nil
	}
	v := this.topNode.val
	this.topNode = this.topNode.next
	return v
}

func (this *LinkedListStack) Top() interface{} {
	if this.IsEmpty() {
		return nil
	}
	return this.topNode.val
}

func (this *LinkedListStack) Flush() {
	this.topNode = nil
}

func (this *LinkedListStack) Print() {
	if this.IsEmpty() {
		fmt.Println("empty stack")
	} else {
		cur := this.topNode
		for nil != cur {
			fmt.Println(cur.val)
			cur = cur.next
		}
	}
}
